### PR TITLE
Write AppsV1DaemonSetStatus test - +3 endpoints

### LINKED ISSUE #100437: Write AppsV1DaemonSetStatus test - +3 endpoints
# Progress <code>[4/6]</code>

-   [x] APISnoop org-flow : [AppsV1 DaemonSet Status lifecycle Test - +3](https://github.com/apisnoop/ticket-writing/blob/create-daemonset-status-lifecycle-test/Appsv1DaemonSetStatusLifecycleTest.org)
-   [x] test approval issue : #100437
-   [X] test pr : #100507
-   [X] two weeks soak start date : [Testgrid Link](https://testgrid.k8s.io/sig-release-master-informing#gce-cos-master-serial&include-filter-by-regex=should%20verify%20changes%20to%20a%20daemon%20set%20status&graph-metrics=test-duration-minutes&width=5) 26 May
-   [ ] two weeks soak end date : 9 June
-   [ ] test promotion pr : kubernetes/kubernetes #102348

# Identifying an untested feature Using APISnoop

According to this APIsnoop query, there are still some remaining DaemonSet endpoints which are untested.

```sql-mode
    SELECT
      endpoint,
      path,
      kind
      FROM testing.untested_stable_endpoint
      where eligible is true
      and endpoint ilike '%DaemonSetStatus'
      order by kind, endpoint desc
      limit 10;
```

```example
                  endpoint                |                             path                              |   kind
  ----------------------------------------+---------------------------------------------------------------+-----------
   replaceAppsV1NamespacedDaemonSetStatus | /apis/apps/v1/namespaces/{namespace}/daemonsets/{name}/status | DaemonSet
   readAppsV1NamespacedDaemonSetStatus    | /apis/apps/v1/namespaces/{namespace}/daemonsets/{name}/status | DaemonSet
   patchAppsV1NamespacedDaemonSetStatus   | /apis/apps/v1/namespaces/{namespace}/daemonsets/{name}/status | DaemonSet
  (3 rows)

```

# API Reference and feature documentation

-   [Kubernetes API Reference Docs](https://kubernetes.io/docs/reference/kubernetes-api/)
-   [Kubernetes API / Workload Resources / DaemonSet](https://kubernetes.io/docs/reference/kubernetes-api/workloads-resources/daemon-set-v1/)
-   [client-go - DaemonSet](https://github.com/kubernetes/client-go/blob/master/kubernetes/typed/apps/v1/daemonset.go)

# The mock test

## Test outline

1.  Create a watch to track daemon set events that DaemonSet with a static label

2.  Create a daemon set with a static label. Confirm that the pods are running.

3.  Get the daemon set status. Parse the response and confirm that the daemon set status conditions can be listed.

4.  Update the daemon set status. Confirm via the watch that the status has been updated.

5.  Patch the daemon set status. Confirm via the watch that the status has been patched.

## Test the functionality in Go

Using an existing [conformance test](https://github.com/kubernetes/kubernetes/blob/cf3374e43491c594071548f68d4357fc9ed537ea/test/e2e/apps/daemon_set.go#L155-L175) as a template for a new [ginkgo test](https://github.com/ii/kubernetes/blob/ca3aa6f5af1b545b116b52c717b866e43c79079b/test/e2e/apps/daemon_set.go#L812-L947) which validates that three new endpoints are hit.

# Verifying increase in coverage with APISnoop

## Listing endpoints hit by the new e2e test

This query shows the endpoints hit within a short period of running the e2e test

```sql-mode
select distinct  endpoint, right(useragent,75) AS useragent
from testing.audit_event
where endpoint ilike '%DaemonSetStatus%'
and release_date::BIGINT > round(((EXTRACT(EPOCH FROM NOW()))::numeric)*1000,0) - 60000
and useragent like 'e2e%'
order by endpoint
limit 10;
```

```example
                endpoint                |                                  useragent
----------------------------------------+-----------------------------------------------------------------------------
 patchAppsV1NamespacedDaemonSetStatus   | [sig-apps] Daemon set [Serial] should verify changes to a daemon set status
 readAppsV1NamespacedDaemonSetStatus    | [sig-apps] Daemon set [Serial] should verify changes to a daemon set status
 replaceAppsV1NamespacedDaemonSetStatus | [sig-apps] Daemon set [Serial] should verify changes to a daemon set status
(3 rows)

```

# Final notes

If a test with these calls gets merged, **test coverage will go up by 3 points**

This test is also created with the goal of conformance promotion.

---

/sig testing
/sig architecture
/area conformance

### LINKED ISSUE #100486: Write ApiService Test - +2 endpoints
# Progress <code>[3/6]</code>

- [x] APISnoop org-flow: [APIServiceTest.org](https://github.com/apisnoop/ticket-writing/blob/master/APIServiceTest.org)
- [x] Test approval issue: [#100486](https://github.com/kubernetes/kubernetes/issues/100486)
- [x] Test PR: [#103092](https://github.com/kubernetes/kubernetes/pull/103092)
- [ ] Two weeks soak start date: testgrid-link
- [ ] Two weeks soak end date:
- [ ] Test promotion PR: kubernetes/kubernetes#?

# Identifying an untested feature Using APISnoop

According to this APIsnoop query, there are still some remaining APIService endpoints which are untested.

```sql-mode
    SELECT
      endpoint,
      path,
      kind
      FROM testing.untested_stable_endpoint
      where eligible is true
      and endpoint ilike '%APIService%'
      order by kind, endpoint desc
      limit 10;
```

```example
                    endpoint                   |                           path                            |    kind
  ---------------------------------------------+-----------------------------------------------------------+------------
   replaceApiregistrationV1APIServiceStatus    | /apis/apiregistration.k8s.io/v1/apiservices/{name}/status | APIService
   replaceApiregistrationV1APIService          | /apis/apiregistration.k8s.io/v1/apiservices/{name}        | APIService
   patchApiregistrationV1APIServiceStatus      | /apis/apiregistration.k8s.io/v1/apiservices/{name}/status | APIService
   deleteApiregistrationV1CollectionAPIService | /apis/apiregistration.k8s.io/v1/apiservices               | APIService
  (4 rows)

```

This test will deal with the following endpoints;

- [replaceApiregistrationV1APIServiceStatus](https://kubernetes.io/docs/reference/kubernetes-api/cluster-resources/api-service-v1/#update-replace-status-of-the-specified-apiservice)
- [deleteApiregistrationV1CollectionAPIService](https://kubernetes.io/docs/reference/kubernetes-api/cluster-resources/api-service-v1/#deletecollection-delete-collection-of-apiservice)

# API Reference and feature documentation

- [Kubernetes API Reference Docs](https://kubernetes.io/docs/reference/kubernetes-api/)
- [Kubernetes API / Cluster Resources / APIService](https://kubernetes.io/docs/reference/kubernetes-api/cluster-resources/api-service-v1/)
- [client-go](https://github.com/kubernetes/client-go/blob/master/kubernetes/typed)

# The mock test

## Test outline

This test uses the current APIService test as a template to create a new APIService, v1beta1.wardle.example.com before testing the endpoints.

1.  Create a watch and list the APIServices with a label selector.

2.  Update the APIService Status with a new set of status conditions. Confirm that the conditions are found via the watch.

3.  Use `DeleteCollection` and a label selector to delete the APIService. Confirm that the deletion was successful.

## Test the functionality in Go

Using the existing [conformance test](https://github.com/kubernetes/kubernetes/blob/2495ec7f1152394dbf096976211f37b21a3e232d/test/e2e/apimachinery/aggregator.go#L99-L102) as a template to create a new [ginkgo test](https://github.com/ii/kubernetes/blob/create-apiservice-test/test/e2e/apimachinery/aggregator.go#L869-L936) which validates that the two endpoints are hit.

## Log sample

    [It] Should be able to replace APIServiceStatus and delete a collection of APIServices
      /home/ii/go/src/k8s.io/kubernetes/test/e2e/apimachinery/aggregator.go:112
    STEP: Registering the sample API server.
    Jun 22 14:03:27.530: INFO: new replicaset for deployment "beta-sample-apiserver-deployment" is yet to be created
    Jun 22 14:03:29.598: INFO: deployment status: v1.DeploymentStatus{ObservedGeneration:1, Replicas:1, UpdatedReplicas:1, ReadyReplicas:0, AvailableReplicas:0, UnavailableReplias:1, Conditions:[]v1.DeploymentCondition{v1.DeploymentCondition{Type:"Available", Status:"False", LastUpdateTime:v1.Time{Time:time.Time{wall:0x0, ext:63759924207, loc:(*time.Location)(0x9510540)}}, LastTransitionTime:v1.Time{Time:time.Time{wall:0x0, ext:63759924207, loc:(*time.Location)(0x9510540)}}, Reason:"MinimumReplicasUnavailable", Message:"Deployment does not have minimum availability."}, v1.DeploymentCondition{Type:"Progressing", Status:"True", LastUpdateTime:v1.Time{Time:time.Time{wall:0x0, ext:63759924207,loc:(*time.Location)(0x9510540)}}, LastTransitionTime:v1.Time{Time:time.Time{wall:0x0, ext:63759924207, loc:(*time.Location)(0x9510540)}}, Reason:"ReplicaSetUpdated", Message:"ReplicaSet \"beta-sample-apiserver-deployment-f8568d796\" is progressing."}}, CollisionCount:(*int32)(nil)}
    Jun 22 14:03:32.746: INFO: Waited 1.12209567s for the sample-apiserver to be ready to handle requests.
    STEP: updating the APIService Status
    Jun 22 14:03:32.761: INFO: updatedStatus.Conditions: []v1.APIServiceCondition{v1.APIServiceCondition{Type:"Available", Status:"True", LastTransitionTime:v1.Time{Time:time.Time{wall:0x0, ext:63759924211, loc:(*time.Location)(0x9510540)}}, Reason:"Passed", Message:"all checks passed"}, v1.APIServiceCondition{Type:"StatusUpdate", Status:"True", LastTransitionTime:v1.Time{Time:time.Time{wall:0x0, ext:0, loc:(*time.Location)(nil)}}, Reason:"E2E", Message:"Set from e2e test"}}
    STEP: watching for the APIService to be updated
    Jun 22 14:03:32.763: INFO: Observed APIService v1beta1.wardle.example.com with Labels: map[apiservice:created] & Conditions: {Available True 2021-06-22 14:03:31 +1200 NZST Passed all checks passed}
    Jun 22 14:03:32.763: INFO: Found APIService v1beta1.wardle.example.com with Labels: map[apiservice:created] & Conditions: {StatusUpdate True 0001-01-01 00:00:00 +0000 UTC E2E Set from e2e test}
    Jun 22 14:03:32.763: INFO: APIService Status for v1beta1.wardle.example.com has been updated
    STEP: Delete a collection of APIServices
    Jun 22 14:03:32.823: INFO: APIService v1beta1.wardle.example.com has been deleted.

# Verifying increase in coverage with APISnoop

This query shows all APIService endpoints hit within a short period of running the e2e test, including the two target endpoints.

```sql-mode
select distinct  endpoint, right(useragent,81) AS useragent
from testing.audit_event
where endpoint ilike '%APIService%'
and release_date::BIGINT > round(((EXTRACT(EPOCH FROM NOW()))::numeric)*1000,0) - 60000
and useragent like 'e2e%'
order by endpoint
limit 10;
```

```example
                  endpoint                   |                                     useragent
---------------------------------------------+-----------------------------------------------------------------------------------
 createApiregistrationV1APIService           | Should be able to replace APIServiceStatus and delete a collection of APIServices
 deleteApiregistrationV1APIService           | Should be able to replace APIServiceStatus and delete a collection of APIServices
 deleteApiregistrationV1CollectionAPIService | Should be able to replace APIServiceStatus and delete a collection of APIServices
 listApiregistrationV1APIService             | Should be able to replace APIServiceStatus and delete a collection of APIServices
 readApiregistrationV1APIService             | Should be able to replace APIServiceStatus and delete a collection of APIServices
 replaceApiregistrationV1APIServiceStatus    | Should be able to replace APIServiceStatus and delete a collection of APIServices
(6 rows)

```

# Final notes

If a test with these calls gets merged, **test coverage will go up by 2 points**

This test is also created with the goal of conformance promotion.

---

/sig testing
/sig architecture
/area conformance

### PR DESCRIPTION
**Pull request by riaankleinhans**
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it:**
This PR adds a test to test the following untested endpoints:
- replaceAppsV1NamespacedDaemonSetStatus
- readAppsV1NamespacedDaemonSetStatus
- patchAppsV1NamespacedDaemonSetStatus  

**Which issue(s) this PR fixes:**
Fixes #100437
**Testgrid Link:**
[Testgrid Link](https://testgrid.k8s.io/sig-release-master-informing#gce-cos-master-serial&include-filter-by-regex=should%20verify%20changes%20to%20a%20daemon%20set%20status&graph-metrics=test-duration-minutes&width=5)

**Special notes for your reviewer:**
Adds +3 endpoint test coverage (good for conformance)

**Does this PR introduce a user-facing change?:**
```
NONE

```

**Release note:**
```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:**
```
NONE

```

/sig testing
/sig architecture
/area conformance

### CODE REVIEW

**Comment by Riaankl** [comment]
/triage accepted
Discussed in Conformance meeting 23 March 2021
PR for 1.22

**Comment by Riaankl** [comment]
/hold
Can merge to 1.22 after Thaw on 8 April 2021

**Comment by spiffxp** [comment]
/hold cancel
/milestone v1.22
Going to use a milestone to signify this isn't intended for v1.21

File: test/e2e/apps/daemon_set.go
Comment (smarterclayton) [inline test/e2e/apps/daemon_set.go:934]: minor style thing, you don't need an else here.  If you "return" / "continue" / "break" in an if, you should exclude the else for clarity.

Comment (heyste) [reply test/e2e/apps/daemon_set.go:934]: I did try that but found [verify-staticcheck.sh](https://github.com/kubernetes/kubernetes/blob/master/hack/verify-staticcheck.sh) fails with `the surrounding loop is unconditionally terminated (SA4004)`. Using `else` was my way of dealing with that error.

**Review by ravisantoshgudimetla — APPROVED** [review]
Thank you for working on this @Riaankl @heyste 

/lgtm

**Comment by Riaankl** [comment]
/assign @soltysh @mattfarina
This e2e test is to clean up the last few DaemonSet endpoints.
Following the same pattern as previous Mergde Conformance tests
It has been reviewed, can we please get a `/approve` to get it on the test grid.

File: test/e2e/apps/daemon_set.go
Comment (heyste) [inline test/e2e/apps/daemon_set.go:895]: @ravisantoshgudimetla Thanks for the earlier review. I've updated this section as there was a potential bug where I had an extra `return`. Also, line 893-4 includes the object for the event now.

Can I get another lgtm? TIA

File: test/e2e/apps/daemon_set.go
Comment (heyste) [inline test/e2e/apps/daemon_set.go:938]: @ravisantoshgudimetla same as line 889-95

File: test/e2e/apps/daemon_set.go
Comment (smarterclayton) [reply test/e2e/apps/daemon_set.go]: I don't think the static check would be right in that case anyway.  Static check doesn't trump style.

**Comment by smarterclayton** [comment]
Changes look ok, once ravi provides lgtm I will approve (if this is sufficiently reliable, which it seems like it is?)

File: test/e2e/apps/daemon_set.go
Comment (heyste) [reply test/e2e/apps/daemon_set.go]: I updated this yesterday. The else block has been removed.

**Comment by heyste** [comment]
/test pull-kubernetes-node-e2e-containerd

x2 flakes
- e2e.go: Node Tests
```
error during go run /go/src/k8s.io/kubernetes/test/e2e_node/runner/remote/run_remote.go --cleanup --logtostderr --vmodule=*=4 --ssh-env=gce --results-dir=/workspace/_artifacts --project=k8s-infra-e2e-boskos-006 --zone=us-west1-b --ssh-user=prow --ssh-key=/workspace/.ssh/google_compute_engine --ginkgo-flags=--nodes=8 --focus="\[NodeConformance\]" --skip="\[Flaky\]|\[Slow\]|\[Serial\]" --test_args=--container-runtime=remote --container-runtime-endpoint=unix:///run/containerd/containerd.sock --container-runtime-process-name=/usr/bin/containerd --container-runtime-pid-file= --kubelet-flags="--cgroups-per-qos=true --cgroup-root=/ --runtime-cgroups=/system.slice/containerd.service" --extra-log="{\"name\": \"containerd.log\", \"journalctl\": [\"-u\", \"containerd*\"]}" --test-timeout=1h5m0s --image-config-file=/workspace/test-infra/jobs/e2e_node/containerd/image-config.yaml: exit status 1
```
- E2eNode  Suite: [sig-node] Summary API [NodeConformance] when querying  /stats/summary should report resource usage through the stats api
```
_output/local/go/src/k8s.io/kubernetes/test/e2e_node/summary_test.go:54
Failed after 22.319s.
Expected
    <string>: Summary
to match fields: {
.Node.SystemContainers[pods].CPU:
	Expected
	    <string>: CPUStats
	to match fields: {
	.UsageNanoCores:
		Expected
		    <*uint64 | 0xc000862fa8>: 0
		To satisfy at least one of these matchers: [%!s(*matchers.BeZeroMatcher=&{}) %!s(*gstruct.PointerMatcher=&{0xc001028f60 Expected
		    <uint64>: 0
		to be >=
		    <int>: 10000})]
	}
	
}
```

**Comment by heyste** [comment]
/test pull-kubernetes-e2e-gce-ubuntu-containerd

x2 flakes

- Kubernetes e2e suite: [sig-instrumentation] MetricsGrabber should grab all metrics from a Scheduler.
```
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/instrumentation/monitoring/metrics_grabber.go:81
May 21 00:38:45.977: Unexpected error:
    <*errors.errorString | 0xc003aefe80>: {
        s: "error waiting for scheduler pod to expose metrics: timed out waiting for the condition; the server is currently unable to handle the request (get pods metrics-proxy:10259)",
    }
    error waiting for scheduler pod to expose metrics: timed out waiting for the condition; the server is currently unable to handle the request (get pods metrics-proxy:10259)
occurred
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/instrumentation/monitoring/metrics_grabber.go:88
```
- e2e.go
```
error during ./hack/ginkgo-e2e.sh --ginkgo.skip=\[Slow\]|\[Serial\]|\[Disruptive\]|\[Flaky\]|\[Feature:.+\] --minStartupPods=8 --report-dir=/workspace/_artifacts --disable-log-dump=true: exit status 1
```

**Review by ravisantoshgudimetla — APPROVED** [review]
/lgtm

**Comment by Riaankl** [comment]
/test pull-kubernetes-e2e-gce-ubuntu-containerd

Unrelated flake
```
I0524 19:19:06.053] • Failure [70.200 seconds]
I0524 19:19:06.053] [sig-instrumentation] MetricsGrabber
I0524 19:19:06.053] /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/instrumentation/common/framework.go:23
I0524 19:19:06.053]   should grab all metrics from a Scheduler. [It]
I0524 19:19:06.053]   /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/instrumentation/monitoring/metrics_grabber.go:81
I0524 19:19:06.053] 
I0524 19:19:06.053]   May 24 19:18:56.577: Unexpected error:
```

**Pull request merged**
